### DISCUSSION TITLE
Turning off the cursor requires visible output?

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#17623] [General]
If I do this,

```
	printf("\x1b[?25l"); // cursor off
	Sleep(5000);
	printf("x");
	Sleep(5000);
```
the cursor is on during the first Sleep and off during the second Sleep.  What's up with that?  Is there any kludge that will turn off the cursor immediately without producing any visible output?

**Comment by vefatica** [comment]
That was in WT.  In a console, the cursor is off for the whole 10 seconds.

**Comment by j4james** [comment]
Yeah, this is a bug - the cursor should turn off immediately. I think it's another regression from PR #17194. The `VtEngine` doesn't see the cursor visibility change as something it needs to handle, so it skips painting that frame.

This is a bit messy to fix because the `VtEngine` doesn't have access to that state. But if we added a `cursorVisible` parameter to the `StartPaint` method, and then made the renderer pass in the `_currentCursorOptions.isVisible` flag, we could compare that against the `_lastCursorIsVisible` field to detect whether the visibility has changed. If it has, we shouldn't abort the paint.

I think the check would need to be added to the `XtermEngine::StartPaint` method where it's testing `hr != S_OK`.

https://github.com/microsoft/terminal/blob/295cd17b028d288ff81445f532d9301b44f6ffd9/src/renderer/vt/XtermEngine.cpp#L40-L41

That test would need to be extended to something like this:
```cpp
if (hr != S_OK && (int)cursorVisible == (int)_lastCursorIsVisible)
```

However, if we're going to be merging PR #17510 soon, there may not be a need for this fix, unless it's something we want to backport to version 1.21.

**Reply by vefatica** [reply]
OK, thanks for the info @j4james .  Is that merge likely to happen in the next preview?  I try to use the latest preview though I don't check for a new one as often as I used to.  I just want the cursor off in transient/pipe instances.  I have a couple ways to do that (plugin DLL, startup script) and the results have never been good.  As you probably know, SetConsoleCursorInfo with bVisible=FALSE is no better.

What about a kludge?  Apparently writing something to the screen gets someone's attention.  Hmmm!  I wonder about writing an nbsp (0xA0) followed by a cursor_left_one_column ... something to try.

**Reply by j4james** [reply]
I don't have any insider knowledge regarding the release schedule, but I think there's a good chance it'll be included in the next preview. There's actually already a version of that PR in the nightly canary build if you want to try that out.

**Reply by lhecker** [reply]
We're currently planning to release the next Preview version (one that includes #17510) at the end of August.
